File: IndividualAssignment1/Python/test/benchmarks.py
import os, time, random, psutil
from src.matrix import multiply

def create_matrix(n):
    return [[random.random() for _ in range(n)] for _ in range(n)]

def run_benchmark(n, repetitions):
    process = psutil.Process(os.getpid())
    A = create_matrix(n)
    B = create_matrix(n)

    total_time = total_cpu = 0.0
    peak_mem = 0.0

    for r in range(repetitions):
        mem_before = process.memory_info().rss / (1024 ** 2)
        cpu_before = process.cpu_times().user
        start = time.perf_counter()

        multiply(A, B, n)

        end = time.perf_counter()
        cpu_after = process.cpu_times().user
        mem_after = process.memory_info().rss / (1024 ** 2)

        elapsed = end - start
        cpu_used = cpu_after - cpu_before
        mem_used = mem_after - mem_before

        total_time += elapsed
        total_cpu += cpu_used
        peak_mem = max(peak_mem, mem_used)

        print(f"Run {r+1} - time={elapsed:.4f}s mem={mem_used:.2f}MB cpu={cpu_used:.4f}s")

    print(f"Python summary - n={n} avg={total_time/repetitions:.6f}s peak_mem={peak_mem:.2f}MB avg_cpu={total_cpu/repetitions:.4f}s\n")

def main():
    for size in [128, 256, 512]:
        run_benchmark(size, 3)

if __name__ == "__main__":
    main()

File: IndividualAssignment1/C/test/benchmarks.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <windows.h>
#include <psapi.h>
#include "matrix.c"

static double **createMatrix(int n) {
    double **m = malloc(n * sizeof(double *));
    for (int i = 0; i < n; ++i)
        m[i] = malloc(n * sizeof(double));
    return m;
}

static void releaseMatrix(double **m, int n) {
    for (int i = 0; i < n; ++i)
        free(m[i]);
    free(m);
}

static void fillRandom(double **m, int n) {
    for (int i = 0; i < n; ++i)
        for (int j = 0; j < n; ++j)
            m[i][j] = (double)rand() / RAND_MAX;
}

static double currentMemoryMB() {
    PROCESS_MEMORY_COUNTERS_EX pmc;
    GetProcessMemoryInfo(GetCurrentProcess(), (PROCESS_MEMORY_COUNTERS *)&pmc, sizeof(pmc));
    return pmc.WorkingSetSize / (1024.0 * 1024.0);
}

static double currentCPUTime() {
    FILETIME c, e, k, u;
    GetProcessTimes(GetCurrentProcess(), &c, &e, &k, &u);
    ULARGE_INTEGER t;
    t.LowPart = u.dwLowDateTime;
    t.HighPart = u.dwHighDateTime;
    return t.QuadPart / 1e7;
}

static double benchmark(int n, int runs) {
    double **A = createMatrix(n);
    double **B = createMatrix(n);
    double **C = createMatrix(n);
    fillRandom(A, n);
    fillRandom(B, n);

    LARGE_INTEGER freq, start, stop;
    QueryPerformanceFrequency(&freq);

    double totalTime = 0.0, maxMem = 0.0;
    for (int r = 0; r < runs; ++r) {
        double cpuStart = currentCPUTime();
        QueryPerformanceCounter(&start);
        multiply(A, B, C, n);
        QueryPerformanceCounter(&stop);
        double cpuEnd = currentCPUTime();

        double elapsed = (double)(stop.QuadPart - start.QuadPart) / freq.QuadPart;
        double memNow = currentMemoryMB();
        if (memNow > maxMem) maxMem = memNow;

        totalTime += elapsed;
        printf("Run %d - time=%.4fs mem=%.2fMB cpu=%.4fs\n", r + 1, elapsed, memNow, cpuEnd - cpuStart);
    }

    printf("C summary - n=%d avg=%.6fs peak_mem=%.2fMB\n\n", n, totalTime / runs, maxMem);
    releaseMatrix(A, n);
    releaseMatrix(B, n);
    releaseMatrix(C, n);
    return totalTime / runs;
}

int main() {
    srand((unsigned)time(NULL));
    int sizes[] = {128, 256, 512};
    for (int i = 0; i < 3; ++i)
        benchmark(sizes[i], 3);
    return 0;
}
